File: elixir/lib/symphony/workspace.ex
defmodule Symphony.Workspace do
  @moduledoc """
  Manages per-issue workspace directories.
  Creates isolated dirs, runs hooks, prevents symlink escapes.
  """
  require Logger

  @type hook_config :: %{optional(String.t()) => String.t()}

  @doc "Create a workspace directory for an issue, run after_create hook."
  def create(issue, opts \\ []) do
    root = Keyword.get(opts, :root, Symphony.Config.workspace_root())
    hooks = Keyword.get(opts, :hooks, %{})
    hook_timeout = Keyword.get(opts, :hook_timeout, 60_000)

    identifier = Symphony.GitHub.Issue.identifier(issue)
    dir = Path.join(root, "issue-#{identifier}")

    with :ok <- File.mkdir_p(dir),
         :ok <- validate_path(dir, root),
         :ok <- run_hook(hooks, "after_create", dir, issue, hook_timeout) do
      {:ok, dir}
    end
  end

  @doc "Remove a workspace directory, running before_remove hook first."
  def remove(dir, opts \\ []) do
    hooks = Keyword.get(opts, :hooks, %{})
    hook_timeout = Keyword.get(opts, :hook_timeout, 60_000)
    root = Keyword.get(opts, :root, Symphony.Config.workspace_root())

    with :ok <- validate_path(dir, root),
         :ok <- run_hook(hooks, "before_remove", dir, nil, hook_timeout) do
      File.rm_rf(dir)
      :ok
    end
  end

  @doc "Run the before_run hook."
  def before_run(dir, issue, opts \\ []) do
    hooks = Keyword.get(opts, :hooks, %{})
    hook_timeout = Keyword.get(opts, :hook_timeout, 60_000)
    run_hook(hooks, "before_run", dir, issue, hook_timeout)
  end

  @doc "Run the after_run hook."
  def after_run(dir, issue, opts \\ []) do
    hooks = Keyword.get(opts, :hooks, %{})
    hook_timeout = Keyword.get(opts, :hook_timeout, 60_000)
    run_hook(hooks, "after_run", dir, issue, hook_timeout)
  end

  @doc "List existing workspace directories."
  def list(root \\ nil) do
    root = root || Symphony.Config.workspace_root()

    case File.ls(root) do
      {:ok, entries} ->
        entries
        |> Enum.filter(&String.starts_with?(&1, "issue-"))
        |> Enum.map(&Path.join(root, &1))

      {:error, :enoent} ->
        []
    end
  end

  # Prevent symlink escape
  defp validate_path(dir, root) do
    real_dir = dir |> Path.expand() |> resolve_symlinks()
    real_root = root |> Path.expand() |> resolve_symlinks()

    if String.starts_with?(real_dir, real_root) do
      :ok
    else
      {:error, {:symlink_escape, dir}}
    end
  end

  defp resolve_symlinks(path) do
    case :file.read_link(String.to_charlist(path)) do
      {:ok, target} ->
        target
        |> List.to_string()
        |> Path.expand(Path.dirname(path))
        |> resolve_symlinks()

      {:error, _} ->
        path
    end
  end

  defp run_hook(hooks, name, dir, issue, timeout) do
    case Map.get(hooks, name) do
      nil ->
        :ok

      cmd_template ->
        cmd = render_hook_cmd(cmd_template, issue)
        Logger.info("Running hook #{name}: #{cmd}")

        task =
          Task.async(fn ->
            System.cmd("sh", ["-c", cmd], cd: dir, stderr_to_stdout: true)
          end)

        case Task.yield(task, timeout) || Task.shutdown(task, :brutal_kill) do
          {:ok, {_output, 0}} ->
            :ok

          {:ok, {output, code}} ->
            Logger.error("Hook #{name} failed (exit #{code}): #{output}")
            {:error, {:hook_failed, name, code}}

          nil ->
            Logger.error("Hook #{name} timed out after #{timeout}ms")
            {:error, {:hook_timeout, name}}
        end
    end
  end

  defp render_hook_cmd(template, nil), do: template

  defp render_hook_cmd(template, issue) do
    template
    |> String.replace("{{identifier}}", Symphony.GitHub.Issue.identifier(issue))
    |> String.replace("{{title}}", issue.title || "")
    |> String.replace("{{number}}", to_string(issue.number))
  end
end

File: src/workspace.ts
import { Effect, Data } from "effect"
import { mkdirSync, existsSync, rmSync, realpathSync, statSync } from "node:fs"
import { join } from "node:path"
import { Config } from "./config.js"
import { type Issue, identifier } from "./github/issue.js"

export class WorkspaceError extends Data.TaggedError("WorkspaceError")<{
  readonly reason: string
}> {}

interface HookConfig {
  readonly [name: string]: string | undefined
}

export interface WorkspaceResult {
  readonly path: string
  readonly createdNow: boolean
}

/**
 * Ensure a workspace directory exists for the given issue.
 * Workspaces persist across runs — they are NOT auto-deleted on success.
 * Use `removeWorkspace` for explicit terminal-state cleanup.
 */
export function ensureWorkspace(
  issue: Issue,
  hooks: HookConfig = {},
  hookTimeoutMs = 60_000,
): Effect.Effect<WorkspaceResult, WorkspaceError, Config> {
  return Effect.flatMap(Config, (config) => {
    const wsKey = sanitizeKey(identifier(issue))
    const dir = join(config.workspaceRoot, wsKey)

    return Effect.gen(function* () {
      // Ensure workspace root exists
      yield* Effect.try({
        try: () => mkdirSync(config.workspaceRoot, { recursive: true }),
        catch: (err) => new WorkspaceError({ reason: `mkdir root failed: ${err}` }),
      })

      const createdNow = !existsSync(dir) || !statSync(dir).isDirectory()

      if (createdNow && config.trackerRepo && !config.dryRun) {
        // Clone the tracker repo into the workspace directory
        yield* cloneRepo(config.trackerRepo, dir, hookTimeoutMs)
      } else if (createdNow) {
        yield* Effect.try({
          try: () => mkdirSync(dir, { recursive: true }),
          catch: (err) => new WorkspaceError({ reason: `mkdir failed: ${err}` }),
        })
      }

      yield* validatePath(dir, config.workspaceRoot)

      // after_create runs ONLY when the directory was just created
      if (createdNow) {
        yield* runHook(hooks, "after_create", dir, issue, hookTimeoutMs)
      }

      return { path: dir, createdNow } satisfies WorkspaceResult
    })
  })
}

/** Run before_run hook. */
export function beforeRun(
  dir: string,
  issue: Issue,
  hooks: HookConfig = {},
  hookTimeoutMs = 60_000,
): Effect.Effect<void, WorkspaceError> {
  return runHook(hooks, "before_run", dir, issue, hookTimeoutMs)
}

/** Run after_run hook (best-effort: failures are logged and ignored). */
export function afterRun(
  dir: string,
  issue: Issue,
  hooks: HookConfig = {},
  hookTimeoutMs = 60_000,
): Effect.Effect<void> {
  return runHook(hooks, "after_run", dir, issue, hookTimeoutMs).pipe(
    Effect.catchAll((err) =>
      Effect.logWarning(`after_run hook failed (ignored): ${err}`),
    ),
  )
}

/** Remove a workspace directory (for terminal-state cleanup). */
export function removeWorkspace(
  dir: string,
  hooks: HookConfig = {},
  hookTimeoutMs = 60_000,
): Effect.Effect<void> {
  return Effect.gen(function* () {
    if (!existsSync(dir)) return

    // before_remove failures are logged and ignored
    yield* runHook(hooks, "before_remove", dir, null, hookTimeoutMs).pipe(
      Effect.catchAll((err) =>
        Effect.logWarning(`before_remove hook failed (ignored): ${err}`),
      ),
    )

    yield* Effect.try({
      try: () => rmSync(dir, { recursive: true, force: true }),
      catch: () => void 0,
    }).pipe(Effect.catchAll(() => Effect.void))
  })
}

/** Sanitize identifier: only [A-Za-z0-9._-] allowed, rest replaced with _ */
export function sanitizeKey(id: string): string {
  return id.replace(/[^A-Za-z0-9._-]/g, "_")
}

/** Prevent symlink escape: resolved path must be under root. */
function validatePath(dir: string, root: string): Effect.Effect<void, WorkspaceError> {
  return Effect.try({
    try: () => {
      const realDir = realpathSync(dir)
      const realRoot = realpathSync(root)
      if (!realDir.startsWith(realRoot)) {
        throw new Error(`Symlink escape detected: ${dir} resolves outside ${root}`)
      }
    },
    catch: (err) => new WorkspaceError({ reason: String(err) }),
  })
}

/** Clone a GitHub repo into the target directory via `gh repo clone`. */
function cloneRepo(
  repo: string,
  dir: string,
  timeoutMs: number,
): Effect.Effect<void, WorkspaceError> {
  return Effect.tryPromise({
    try: async () => {
      const proc = Bun.spawn(["gh", "repo", "clone", repo, dir], {
        stdout: "pipe",
        stderr: "pipe",
      })
      const stdout = await new Response(proc.stdout).text()
      const stderr = await new Response(proc.stderr).text()
      const exitCode = await proc.exited
      if (exitCode !== 0) {
        throw new Error(`git clone failed (exit ${exitCode}): ${stderr || stdout}`)
      }
    },
    catch: (err) => new WorkspaceError({ reason: String(err) }),
  }).pipe(
    Effect.timeout(`${timeoutMs} millis`),
    Effect.catchTag("TimeoutException", () =>
      Effect.fail(new WorkspaceError({ reason: `Clone of ${repo} timed out after ${timeoutMs}ms` })),
    ),
    Effect.asVoid,
  )
}

function renderHookCmd(template: string, issue: Issue | null): string {
  if (!issue) return template
  return template
    .replace(/\{\{identifier\}\}/g, identifier(issue))
    .replace(/\{\{title\}\}/g, issue.title)
    .replace(/\{\{number\}\}/g, String(issue.number))
}

function runHook(
  hooks: HookConfig,
  name: string,
  dir: string,
  issue: Issue | null,
  timeoutMs: number,
): Effect.Effect<void, WorkspaceError> {
  const cmdTemplate = hooks[name]
  if (!cmdTemplate) return Effect.void

  const cmd = renderHookCmd(cmdTemplate, issue)

  return Effect.tryPromise({
    try: async () => {
      const proc = Bun.spawn(["sh", "-c", cmd], {
        cwd: dir,
        stdout: "pipe",
        stderr: "pipe",
      })
      const stdout = await new Response(proc.stdout).text()
      const stderr = await new Response(proc.stderr).text()
      const exitCode = await proc.exited
      if (exitCode !== 0) {
        throw new Error(`Hook ${name} failed (exit ${exitCode}): ${stderr || stdout}`)
      }
    },
    catch: (err) => new WorkspaceError({ reason: String(err) }),
  }).pipe(
    Effect.timeout(`${timeoutMs} millis`),
    Effect.catchTag("TimeoutException", () =>
      Effect.fail(new WorkspaceError({ reason: `Hook ${name} timed out after ${timeoutMs}ms` })),
    ),
    Effect.asVoid,
  )
}
